src/Microsoft.AspNetCore.Server.IISIntegration/Server/IISHttpContext.cs

```diff
@@ -258,24 +239,266 @@ private IISAwaitable DoFlushAsync()
         {
             await InitializeResponseAwaited(data.Count);
 
             // WriteAsyncAwaited is only called for the first write to the body.
             // Ensure headers are flushed if Write(Chunked)Async isn't called.
-            await Output.WriteAsync(data, cancellationToken: cancellationToken);
+
+            Task writeTask;
+            lock (_stateSync)
+            {
+                DisableReads();
+
+                // Want to guarantee that data has been written to the pipe before releasing the lock.
+                writeTask = Output.WriteAsync(data, cancellationToken: cancellationToken);
+            }
+            await writeTask;
+        }
+
+        /// <summary>
+        /// Flushes the data in the output pipe
+        /// </summary>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        public async Task FlushAsync(CancellationToken cancellationToken = default(CancellationToken))
+        {
+            await InitializeResponse(0);
+            await Output.FlushAsync(cancellationToken);
+        }
+
+        public void StartProcessingRequestAndResponseBody()
+        {
+            if (_readWriteTask == null)
+            {
+                lock (_createReadWriteBodySync)
+                {
+                    if (_readWriteTask == null)
+                    {
+                        _readWriteTask = ConsumeAsync();
+                    }
+                }
+            }
+        }
+
+        // ConsumeAsync is called when either the first read or first write is done. 
+        // There are two modes for reading and writing to the request/response bodies.
+        // 1. Await all reads and try to read from the Output pipe
+        // 2. Done reading and await all writes.
+        public async Task ConsumeAsync()
+        {
+            await ReadAndWriteLoopAsync();
+
+            // If we are done writing, complete the output pipe and return
+            // Input Pipe will be closed when ReadAndWriteLoopAsync returns
+            if (_doneWriting)
+            {
+                Output.Reader.Complete();
+                return;
+            }
+
+            await WriteLoopAsync();
+        }
+
+        public async Task ReadAndWriteLoopAsync()
+        {
+            // First we check if there is anything to write from the Output pipe
+            // If there is, we call WriteToIISAsync
+            while (!_doneReading)
+            {
+                // Check if Output pipe has anything to write to IIS.
+                if (Output.Reader.TryRead(out var readResult))
+                {
+                    var buffer = readResult.Buffer;
+                    var consumed = buffer.End;
+
+                    try
+                    {
+                        // If the output pipe is termniated, cancel the request.
+                        if (readResult.IsCancelled)
```

Comment:
Do you ever call CancelPendingRead() on the output pipe?

Reply:
No. Should I?